### PR TITLE
[GE] Constraints for Mouse Movement

### PR DESCRIPTION
**Pull request by msDestiny14**
Using the Shift Key will now constrain the mouse movement of GUI to either the X or Y axis.

### CODE REVIEW

File: guiEditor/src/diagram/workbench.tsx [hunk 1]
```diff
@@ -36,6 +36,13 @@ export const isFramePortData = (variableToCheck: any): variableToCheck is FrameP
     } else return false;
 };
 
+export enum ConstraintDirection {
+    NONE = 0,
+    SET = 1,
```
Comment (sebavan) [inline guiEditor/src/diagram/workbench.tsx:41]: Not fully sure what SET is, might be good to add comments on the different values

File: guiEditor/src/diagram/workbench.tsx [hunk 2]
```diff
@@ -44,14 +51,15 @@ export class WorkbenchComponent extends React.Component<IWorkbenchComponentProps
     private _scene: Scene;
     private _selectedGuiNodes: Control[] = [];
     private _ctrlKeyIsPressed = false;
+    private _constraintDirection = ConstraintDirection.NONE;
     private _forcePanning = false;
     private _forceZooming = false;
     private _forceSelecting = false;
     private _outlines = false;
     public _frameIsMoving = false;
```
Comment (sebavan) [inline guiEditor/src/diagram/workbench.tsx:59]: Would be good to group private / public together for readability ?

File: guiEditor/src/diagram/workbench.tsx [hunk 3]
```diff
@@ -44,14 +51,15 @@ export class WorkbenchComponent extends React.Component<IWorkbenchComponentProps
...
+    private _constraintDirection = ConstraintDirection.NONE;
     private _forcePanning = false;
     private _forceZooming = false;
     private _forceSelecting = false;
     private _outlines = false;
     public _frameIsMoving = false;
     public _isLoading = false;
     public isOverGUINode = false;
-    public artBoardBackground : Rectangle;
+    public artBoardBackground: Rectangle;
```
Comment (sebavan) [inline guiEditor/src/diagram/workbench.tsx:62]: This should be Nullable I guess ?

File: guiEditor/src/diagram/workbench.tsx [hunk 4]
```diff
@@ -157,7 +165,9 @@ export class WorkbenchComponent extends React.Component<IWorkbenchComponentProps
 
     ctrlEvent = (evt: KeyboardEvent) => {
         this._ctrlKeyIsPressed = evt.ctrlKey;
-
+        this._constraintDirection = evt.shiftKey ?
```
Comment (sebavan) [inline guiEditor/src/diagram/workbench.tsx:168]: Kind of hard to decipher ??? might be good on several lines ? (or I am getting old ... or both :-) )

File: guiEditor/src/diagram/workbench.tsx [hunk 5]
```diff
@@ -307,6 +317,16 @@ export class WorkbenchComponent extends React.Component<IWorkbenchComponentProps
         let newX = evt.x - startPos.x;
         let newY = evt.y - startPos.y;
 
+        if (this._constraintDirection === ConstraintDirection.SET) {
+            this._constraintDirection = Math.abs(newX) >= Math.abs(newY) ? ConstraintDirection.X : ConstraintDirection.Y;
+        }
+        if (this._constraintDirection == ConstraintDirection.X) {
```
Comment (sebavan) [inline guiEditor/src/diagram/workbench.tsx:323]: [Nit Picking Level 12000] Could use === like 3 lines above for consistency

File: guiEditor/src/diagram/workbench.tsx [hunk 6]
```diff
@@ -307,6 +317,16 @@ export class WorkbenchComponent extends React.Component<IWorkbenchComponentProps
         let newX = evt.x - startPos.x;
         let newY = evt.y - startPos.y;
 
+        if (this._constraintDirection === ConstraintDirection.SET) {
+            this._constraintDirection = Math.abs(newX) >= Math.abs(newY) ? ConstraintDirection.X : ConstraintDirection.Y;
+        }
+        if (this._constraintDirection == ConstraintDirection.X) {
+            newY = 0;
+        }
+        else if (this._constraintDirection == ConstraintDirection.Y) {
```
Comment (sebavan) [inline guiEditor/src/diagram/workbench.tsx:326]: [Nit Picking Level 12000] Could use === like 3 lines above for consistency

**Review by sebavan — COMMENTED** [review]
LGTM just tiny comments

File: dist/preview release/what's new.md [hunk 7]
```diff
@@ -120,6 +120,7 @@
 - Bug fix to panning ([msDestiny14](https://github.com/msDestiny14))
 - Added black bar and scene explorer view ([msDestiny14](https://github.com/msDestiny14))
 - Added navigation hotkeys ([msDestiny14](https://github.com/msDestiny14))
+- Added mouse constraints using the shift key ([msDestiny14](https://github.com/msDestiny14))
```
[the remark this answers is not in the document]
Comment (msDestiny14) [reply dist/preview release/what's new.md:123]: Fair enough. Should I remove all the other lines?

[the remark this answers is not in the document]
Comment (Popov72) [reply dist/preview release/what's new.md:123]: I think you should remove the whole section, but let's wait on @sebavan or @deltakosh comments.

[the remark this answers is not in the document]
Comment (sebavan) [reply dist/preview release/what's new.md:123]: Yup only adding a top level main feature GUI editor line would be perfect.

**Review by Popov72 — APPROVED** [review]

**Review by sebavan — APPROVED** [review]

**Pull request merged**